---
title: "Dabiri R Assignment"
output: html_document
date: "2024-03-11"
---
This line of code imports the data set to be used into R studio.
```{r}
library(httr)

# Define the URLs for the data
snp_position_url <- "https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2024/main/assignments/UNIX_Assignment/snp_position.txt"
fang_et_al_genotypes_url <- "https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2024/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt"

# Read data from URLs
snp_position <- read.table(url(snp_position_url), header = TRUE, sep = "\t")
fang_et_al_genotypes <- read.table(url(fang_et_al_genotypes_url), header = TRUE, sep = "\t")

# Create the output folder if it doesn't exist
output_folder <- "output_folder"
if (!file.exists(output_folder)) {
  dir.create(output_folder)
}
snp_position_file <- file.path(output_folder, "snp_position.txt")
fang_et_al_genotypes_file <- file.path(output_folder, "fang_et_al_genotypes.txt")

# Write the data to files
write.table(snp_position, file = snp_position_file, sep = "\t", row.names = FALSE)
write.table(fang_et_al_genotypes, file = fang_et_al_genotypes_file, sep = "\t", row.names = FALSE)
```

```{r}
#Data Inspection: The following set of code inspects the data 
###This shows the dimension of the dataframe 
dim(fang_et_al_genotypes) 
dim(snp_position)
###This shows the summary of each data set 
colnames(fang_et_al_genotypes)
colnames(snp_position)
sapply(fang_et_al_genotypes, class)
sapply(snp_position, class)
is.data.frame(fang_et_al_genotypes)
is.data.frame(snp_position)
file.info("fang_et_al_genotypes")
file.info("snp_position.txt")
sum(is.na(fang_et_al_genotypes))
sum(is.na(snp_position))
cor(snp_position[sapply(snp_position, is.numeric)])
any(is.na(fang_et_al_genotypes))
any(is.na(snp_position))
any(duplicated(fang_et_al_genotypes))
any(duplicated(snp_position))
```

```{r}
### Data processing 
# This line of code copies out the data based on the group into a new file Maize_data and teosinte_data
maize_data_group <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMMIL", "ZMMLR", "ZMMMR"), ]
teosinte_data_group <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMPBA", "ZMPIL", "ZMPJA"), ]

```

```{r}
# This line of code will arrange the snp_position data frame by SNP_ID
snp_position_Chr <- snp_position[,-c(2,5:15)] 
snp_position_arranged <- arrange(snp_position_Chr, SNP_ID)
```


```{r}
library(dplyr)

# Remove columns 1 to 3 from the maize_data_group and teosinte_data_group data frame
maize_data_group <- select(maize_data_group, -(1:3))
teosinte_data_group <- select(teosinte_data_group, -(1:3))
# The data will be transposed using this line of code
maize_genotypes_trans <- t(maize_data_group)
teosinte_genotypes_trans <- t(teosinte_data_group)
```


```{r}
teosinte_snp_genotype <- cbind(snp_position_arranged, teosinte_genotypes_trans)
# Combining snp_position_arranged with teosinte_genotypes_trans at the 1,1 position
maize_snp_genotype <- cbind(snp_position_arranged, maize_genotypes_trans)
# Combining snp_position_arranged with maize_genotypes_trans at the 1,1 position

# Print the number of rows for each data frame
nrow(maize_genotypes_trans)
nrow(teosinte_genotypes_trans)
nrow(snp_position_arranged)
```

```{r}
###This replaces all the ? data with -
maize_snp_genotype <- maize_snp_genotype %>% mutate_all(function(x) gsub("\\?", "\\-",x))
teosinte_snp_genotype <- teosinte_snp_genotype %>% mutate_all(function(x) gsub("\\?", "\\-",x))

```

```{r}
# Writing out the Chromosome based on desending and ascending  SNPposition for Maize
library(dplyr)
library(readr)
for (i in 1:10) {
  print(paste("Processing Chromosome", i))
  ascending_data <- maize_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(as.numeric(Position))
  
  descending_data <- maize_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(desc(as.numeric(Position)))
  
 
  write_tsv(ascending_data, paste("Maize_chr", i, "ascending.txt", sep = "_"))
  
 
  write_tsv(descending_data, paste("Maize_chr", i, "descending.txt", sep = "_"))
}
  
```

```{r}
#Writing out the Chromosome based on desending and ascending  SNPposition for Teosinte
for (i in 1:10) {
  print(paste("Processing Chromosome", i))
  
  ascending_data <- teosinte_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(as.numeric(Position))
  
  descending_data <- teosinte_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(desc(as.numeric(Position)))
  
  write_tsv(ascending_data, paste("teosinte_chr", i, "ascending.txt", sep = "_"))
  write_tsv(descending_data, paste("teosinte_chr", i, "descending.txt", sep = "_"))
}

```


#Data Visualization
```{r}
#SNP vs Chromosome boxplot 
library(ggplot2)
maize_snp_genotype$Chromosome <- factor(
  maize_snp_genotype$Chromosome,
  levels = as.character(1:10),
  ordered = TRUE
)

ggplot(maize_snp_genotype, aes(x = Chromosome, y = as.numeric(Position), fill = Chromosome)) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "Chromosome", y = "Position", title = "Boxplot by Chromosome for maize ")
  
teosinte_snp_genotype$Chromosome <- factor(
  teosinte_snp_genotype$Chromosome,
  levels = as.character(1:10),
  ordered = TRUE
)

ggplot(teosinte_snp_genotype, aes(x = Chromosome, y = as.numeric(Position), fill = Chromosome)) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "Chromosome", y = "Position", title = "Boxplot by Chromosome for teosinte")
```

#Data Visualization
```{r}
library(ggplot2)

# For maize
maize_snp_genotype$Chromosome <- factor(
  maize_snp_genotype$Chromosome,
  levels = as.character(1:10),
  ordered = TRUE
)

# Count SNPs per chromosome for maize
maize_snp_counts <- data.frame(table(maize_snp_genotype$Chromosome))

# Plot bar graph for maize
ggplot(maize_snp_counts, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Chromosome", y = "Number of SNPs", title = "SNP Distribution per Chromosome for Maize")

# For teosinte
teosinte_snp_genotype$Chromosome <- factor(
  teosinte_snp_genotype$Chromosome,
  levels = as.character(1:10),
  ordered = TRUE
)

# Count SNPs per chromosome for teosinte
teosinte_snp_counts <- data.frame(table(teosinte_snp_genotype$Chromosome))

# Plot bar graph for teosinte
ggplot(teosinte_snp_counts, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Chromosome", y = "Number of SNPs", title = "SNP Distribution per Chromosome for Teosinte")
```


```{r}

#Heterozygous vs
library(dplyr)
library(tidyr)
library(ggplot2)
library(tidyverse)

cleaned_fang <- fang_et_al_genotypes %>%
  select(-JG_OTU) %>%
  pivot_longer(cols = -c(Sample_ID, Group), names_to = "SNP_ID", values_to = "Sequence")

cleaned_fang <- cleaned_fang %>%
  mutate(new_sequence = case_when(
    Sequence %in% c("A/A", "T/T", "C/C", "G/G") ~ "Homozygous",
    Sequence == "?/?" ~ "Missing",
    TRUE ~ "Heterozygous"
  ))

ggplot(cleaned_fang, aes(x = new_sequence, fill = new_sequence)) +
  geom_bar() +
  theme_bw() +
  labs(title = "Distribution of Genotypes", x = "Genotype", y = "Count")
```

#My own Visualization
```{r}
library(ggplot2)

library(ggplot2)

# library(ggplot2)

# library(ggplot2)

# Create the scatter plot
ggplot(data = genotype_counts) +
  geom_point(mapping = aes(x = Position, y = count)) +
  labs(title = "Scatter Plot of Genotype Varieties by Group") +
<
